### ISSUE TITLE
[CIP10] Generalized Authorization of Signers

### ISSUE THREAD

**Issue by nambrot** [celo-org/celo-proposals#76]
This issue is to discuss CIP10

**Comment by nambrot** [comment]
Original discussion https://github.com/celo-org/celo-proposals/pull/17

**Comment by AlexBHarley** [comment]
Just wanting to wrap up this conversation before I nail down the implementation and amend the proposal. It looks like we'll definitely be relaxing the assumption that a signer can only be authorised for one role.

I see three options available when it comes to actually authorising general roles (listed in order of usefulness but also ease of implementation).

1. Support one signer per role
2. Support one signer for Validator, Vote and Attestation roles, multiple signers for other roles
3. Support multiple signers for every role

Option 2. or 3. sounds more desirable, with 3. being more difficult to implement. For example, in `Attestations.sol` we call [accounts.hasAuthorizedAttestationSigner(issuer)](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/identity/Attestations.sol#L672), determining `hasSigner` on chain is difficult because in Solidity you can't determine the length of a mapping without adding another data structure or counter.

**Comment by nambrot** [comment]
Could you do something like having the notion of a primary signer for a role and keep the overhead the same? If not, 2) sounds perfectly reasonanle

**Comment by AlexBHarley** [comment]
The overarching goals of CIP10 are to 

- allow multiple signers per role 
- allow multiple roles per signer
- allow reusing signers for multiple roles.

This is at odds with the current implementation of having a single unique signer specified for each of the vote, validator, attestation and wallet addresses.

Our current API looks as follows for any given signer,

- authorizeXxxSigner
- removeXxxSigner
- xxxSignerToAccount
- getXxxSigner
- hasAuthorizedXxxSigner

The problem I'm having is trying to present a single and consolidated API that's backwards compatible with these methods. In an ideal world we'd have started with this N-M signer-role relationship and could have an API as follows

- authorizeSigner(address signer, string role)
- completeSignerAuthorization(address account)
- authorizeSignerWithSignature(address signer, v, r, s)
- removeSigner(address signer, string role)
- isSigner(address account, address signer, string role) boolean
- getSigners(address account, string role) address[]
- hasSigner(address account, string role) boolean

In smart contract land, getSigner and hasSigner are the particularly difficult ones to represent. 

One idea floated has been to have a "dedicated" signer which would imply two additional mappings being added to Accounts.sol

mapping(string => {started: boolean, completed: boolean}) dedicatedSigners;
mapping(address => mapping(string => {started: boolean, completed: boolean})) authorizedSigners;

This would of course allow the hasSigner and getSigner to return the dedicated signer.

**Comment by nambrot** [comment]
So I would probably modify your last point, to just have `authorizedSigners` keep track of the authorization and rename `dedicatedSigners` to something like `defaultSigners`. Also `defaultSigners` doesn't really need to keep track of started/completed, no? It could just rely on `authorizedSigners`?

Another question would be whether the `authorizeSigner` methods would automatically set `defaultSigner` when not set. I can see cases for either. I think in our current form, many roles want something like a default signer, but long-term that might not be the case in which case we would incur unnecessary storage changes. So actually, thinking about it, i think the new methods we add shouldn't by default set it, but the existing ones for the legacy roles should (since they depend on it)

**Comment by mikereinhart** [comment]
Hi @AlexBHarley, great to see this CIP moving forward! A few questions:

> It looks like we'll definitely be relaxing the assumption that a signer can only be authorised for one role.

To confirm, the scope you are aiming for would be:
1. A signer can be authorized for several roles _for the same account_
2. A signer cannot ever be authorized for two accounts

I remember seeing somewhere that how the protocol handles slashing makes it very difficult to have a signer be authorized for a role with two accounts (ie. voting for two accounts). This would be nice to have, but if it is especially difficult then keeping it out of scope makes sense.

> 2. Support one signer for Validator, Vote and Attestation roles, multiple signers for other roles

Are there any 'other roles' that are known at this time or is this purposefully abstract for variable use cases?

Finally, regarding the overarching goals of the CIP:
> allow multiple signers per role

Do you mean allowing smart contracts as signers or having multiple (EOA) signers recognized by the protocol as authorized for a role? My preference would be smart contracts.

Thanks!

**Comment by mikereinhart** [comment]
We were able to discuss some of my questions during the All Core Dev Call #7. For reference:
- Yes, a signer can be authorized for several roles for the same account, but it cannot sign for two different accounts
- Multiple signers can be authorized for the same role. These signers can be EOA addresses or smart contracts

Excited to see this reach mainnet!

**Comment by nategraf** [comment]
As we move to smart contract accounts as the standard for Valora, and potentially standardize techniques for smart contract accounts more generally, this CIP will become increasingly important.

At the moment, this CIP seems to primarily address on-chain usage. It would be nice to see off-chain usage considered and standardized. For example, CIP-8 needs a standardized method for defining what keys are valid signers, and some dApps require off-chain signatures. An example of this was 0x exchange version 1. It used an off-chain order book signed by the account keys. (Version two used a generalized contract signature standard https://eips.ethereum.org/EIPS/eip-1271)

I believe CIP-10 can be used for off-chain signature verification in the following way:
* Bob has received an ECDSA signature `(v, r, s)` from Alice as account `0xa11ce...`
* Bob recovers from the signature the signer address `0x5197e2...`
* Bob executes view call `Accounts.isSigner(0xa11ce..., 0x5197e2..., "cryptod.app/message/signer")` and checks the response. If `true`, Bob accepts the signature, otherwise rejects.

If used in this manner, it should work quite well because:
A) By virtue of producing a signature, the signer address must be an EOA and not a contract
B) Because users in the (updated) CIP-10 can specify multiple signers per role this does not force them to _only_ user the EOA for a given role
C) It requires the minimum number of contract calls (one) and the response can be cached for a certain amount of time determined by the application's security needs.

What do y'all think about explicitly endorsing such a usage in the CIP?

**Comment by nambrot** [comment]
i 100% agree that it would be very nice to combine CIP8 and 10 in that regard, though given how long it has taken CIP10 to land I'd be wary of expanding scope at this point. 

I suspect that what you are looking for is in effect mostly social consensus that certain applications consider using some kind of generalized schema on top of CIP8 as an optimistic authorization option?

**Comment by nategraf** [comment]
I am indeed seeking social consensus on using CIP-10 for off-chain signatures. I don't think any changes would be needed to the spec itself, although I understand the hesitation to add anything in case it becomes controversial. (One clarification is that I see this as more general than CIP-8, but CIP-8 would definitly be a consumer of off-chain signatures and we could amend CIP-8 with a note about this being the standard and preferred way to verify a signature, as it currently does not provide strong guidance on what keys to use)

**Comment by YazzyYaz** [comment]
Yeah considering CIP-10 is currently in Last Call status with review period expiring in May 4th, combining specifications now seems a bit too late given rough consensus was achieved on the ACD call to move forward to Last Call.

We can still move it back to the Draft status if you feel this specification is incomplete or has security risks or needs a drastic improvement due to some flaw in the spec. Just wondering if it's worth the social cost to do so for the sake of combining specs or if both CIP-10 and CIP-8 as separate specs can still compliment each other.

Happy to move either way, I leave the call on what to do on the technical implementation side to you the core devs :)

**Comment by nategraf** [comment]
To clarify, I'm not suggesting any change to the substance of CIP-10 and am happy to write up another CIP to compliment it specifically for the off-chain signature use case. Also, this is not specific to CIP-8
